Support for more general geometry attributes

panxinmiao opened the pull request:
This pull request adds support for decoding generic geometry attributes (such as `tangents`, `joints`, `weights`, etc) to meet the requirements of specifications such as glTF.
The API remains backward compatible.

In addition, this PR also fixes the `normals` decoding issue mentioned in Issue #54.

panxinmiao commented:
Hi!  The previous commit did not handle point cloud format— this has now been fixed.

Unit tests have also been added.

william-silversmith commented on src/DracoPy.pyx:
```diff
@@ -211,6 +272,31 @@ def encode(
+        assert len(joints.shape) == 2, "Joints must be 2D"
+        joint_channel = joints.shape[1]
+        assert 1 <= joint_channel <= 16, "Number of joint channels must be in range [1, 16]"
+        joints = joints.astype(np.uint16)
```
might want to write `astype(np.uint16, copy=False)` to prevent making unnecessary extra copies

william-silversmith commented on src/DracoPy.pyx:
```diff
@@ -37,42 +70,56 @@ class DracoPointCloud:
-        points_ = self.data_struct['points']
-        N = len(points_) // 3
-        return np.array(points_).reshape((N, 3))
+        position_attr = self.get_attribute_by_type(0)  # POSITION=0
```
Can you point to the documentation where these constants are defined? I looked through the spec but maybe I missed it.

william-silversmith commented on src/DracoPy.h:
```diff
@@ -107,66 +112,106 @@ namespace DracoFunctions {
+      switch (att->data_type()) {
+        case draco::DT_FLOAT32: {
+          attr.float_data.reserve(num_values);
+          float *values = new float[att->num_components()];
```
Good idea to use `std::unique_ptr<float[]> values(new float[att->num_components()]);` for safety. You can still control the time of deallocation with `values.reset()`.

panxinmiao commented:
Hi, I've updated the code to support encoding more robust general attributes—not just `tangent`, `joint`, and `weight`. 
The new API clearly specifies that attributes of type `GENERIC` are identified and distinguished using "unique_id" to avoid ambiguity.

In addition, I've added enum definitions for Draco.AttributeType and Draco.DataType to the Python interface for easier usage.

william-silversmith commented on src/DracoPy.pyx:
```diff
@@ -202,12 +225,11 @@ def encode(
-        vertices. K is the number of joint influences per vertex. Use None if mesh does not have skeletal data.
-    Weights is a numpy array of joint weights (float) with shape (N, K). N is the number of
-        vertices. K is the number of joint influences per vertex. Use None if mesh does not have skeletal data.
+    Generic_attributes is a dictionary of additional attributes to encode.
```
I think it would be good to show a small example of how to use it. The values section is explained well, but I think for someone that doesn't know what it is, unique_ids is not totally self explanatory.

panxinmiao replied:
Updated.

william-silversmith commented:
Thank you! A very valuable improvement.
